libyul/optimiser/BlockHasher.h

```diff
@@ -55,10 +55,15 @@ class HasherBase
 		hash32(static_cast<uint32_t>(_value >> 32));
 	}
 
-
 	uint64_t m_hash = fnvEmptyHash;
 };
 
+class ASTHasherBase: public HasherBase
```

Comment:
I expected this to be a bit simpler, i.e. just a small helper in an unnamed namespace in `BlockHasher.cpp`, but I guess this is fine too if you prefer it that way :)

Reply:
Then I'd have to change the visibility of the member functions :) I thought this was cleaner.

Reply:
No, why? Maybe it's not entirely clear what I meant.

Note that there's already a `namespace { ... }` block at the top of `BlockHasher.cpp` that defines a `compileTimeLiteralHash()` helper. This is used by both implementations and has no issues with their visibility. What I meant was to put the helper there.

Reply:
Right, without using the `hash{8,16,32,64}` helpers it's of course possible:)

Reply:
Ah wait, those are defined in `HasherBase`. I was assuming they were global.

So ok, you do have a point here :)

Reply:
It did itch me a little, something like this would also be possible (essentially the same thing hasher base is doing):

```cpp
template<typename Component>
constexpr void combine(uint64_t& _hash, Component const& _c)
{
	auto const* data = reinterpret_cast<uint8_t const *>(&_c);
	for (size_t i = 0; i < sizeof(Component); ++i)
	{
		_hash *= HasherBase::fnvPrime;
		_hash ^= data[i];
	}
}

template<typename LiteralComponents, auto... Index>
constexpr uint64_t hashComponentsImpl(std::index_sequence<Index...>, LiteralComponents const& _components)
{
	constexpr std::array indices{Index...};
	uint64_t hashValue = HasherBase::fnvEmptyHash;
	(combine(hashValue, std::get<indices[Index]>(_components)),...);
	return hashValue;
}

template<typename... LiteralComponents>
constexpr uint64_t hashComponents(LiteralComponents const&... _components)
{
	return hashComponentsImpl(std::make_index_sequence<sizeof...(LiteralComponents)>(), std::tie(_components...));
}

uint64_t hashLiteral(Literal const& _literal)
{
	if (!_literal.value.unlimited())
		return hashComponents(compileTimeLiteralHash("Literal"), std::hash<u256>{}(_literal.value.value()), _literal.type.hash(), _literal.kind, _literal.value.unlimited());
	else
		return hashComponents(compileTimeLiteralHash("Literal"), std::hash<std::string>{}(_literal.value.builtinStringLiteralValue()), _literal.type.hash(), _literal.kind, _literal.value.unlimited());
}
```